File: src/daytrace/fusion.py
from __future__ import annotations

from collections.abc import Callable, Iterable
from dataclasses import replace

from daytrace.models import (
    ActivitySlice,
    ContextSignal,
    DiagnosticCode,
    SanitizedObservation,
    SourceKind,
)


BROWSER_APPS = ("edge", "chrome", "firefox", "brave", "opera", "safari")
EDITOR_APPS = ("code", "codium", "pycharm", "idea", "sublime", "zed")


def _app_matches(app: str | None, families: tuple[str, ...]) -> bool:
    folded = (app or "").casefold()
    return any(name in folded for name in families)


def _context(item: SanitizedObservation) -> ContextSignal:
    return ContextSignal(
        kind=item.kind,
        evidence_id=item.evidence_id,
        title=item.title,
        project=item.project,
        file=item.file,
        url_host=item.url_host,
        url_path=item.url_path,
        language=item.language,
    )


def _focused_slice(
    start,
    end,
    winner: SanitizedObservation,
    contexts: tuple[ContextSignal, ...],
) -> ActivitySlice:
    evidence = tuple(
        sorted({winner.evidence_id, *(item.evidence_id for item in contexts)})
    )
    return ActivitySlice(
        start,
        end,
        True,
        winner.app,
        winner.title,
        contexts,
        evidence,
    )


def _evidence_only_slice(
    start,
    end,
    covering: Iterable[SanitizedObservation],
) -> ActivitySlice:
    contexts = tuple(
        sorted((_context(item) for item in covering), key=lambda item: item.evidence_id)
    )
    evidence = tuple(item.evidence_id for item in contexts)
    return ActivitySlice(start, end, False, None, None, contexts, evidence)


def _merge_identical_touching(
    slices: tuple[ActivitySlice, ...],
) -> tuple[ActivitySlice, ...]:
    merged: list[ActivitySlice] = []
    for item in slices:
        previous = merged[-1] if merged else None
        if previous and previous.end == item.start and (
            previous.focused,
            previous.app,
            previous.title,
            previous.contexts,
        ) == (item.focused, item.app, item.title, item.contexts):
            merged[-1] = replace(
                previous,
                end=item.end,
                evidence_ids=tuple(
                    sorted({*previous.evidence_ids, *item.evidence_ids})
                ),
            )
        else:
            merged.append(item)
    return tuple(merged)


def fuse_observations(
    observations: Iterable[SanitizedObservation],
    diagnose: Callable[[DiagnosticCode], None],
) -> tuple[ActivitySlice, ...]:
    items = tuple(
        sorted(observations, key=lambda item: (item.start, item.evidence_id))
    )
    windows = tuple(item for item in items if item.kind is SourceKind.WINDOW)
    context_items = tuple(item for item in items if item.kind is not SourceKind.WINDOW)
    boundaries = sorted({point for item in items for point in (item.start, item.end)})
    output: list[ActivitySlice] = []

    for start, end in zip(boundaries, boundaries[1:]):
        covering_windows = [
            item for item in windows if item.start < end and item.end > start
        ]
        if covering_windows:
            if len(covering_windows) > 1:
                diagnose(DiagnosticCode.WINDOW_CONFLICT)
            winner = min(covering_windows, key=lambda item: item.evidence_id)
            contexts: list[ContextSignal] = []
            for item in context_items:
                if item.start >= end or item.end <= start:
                    continue
                attach = (
                    item.kind is SourceKind.BROWSER
                    and _app_matches(winner.app, BROWSER_APPS)
                ) or (
                    item.kind is SourceKind.EDITOR
                    and _app_matches(winner.app, EDITOR_APPS)
                )
                if attach:
                    contexts.append(_context(item))
            output.append(
                _focused_slice(
                    start,
                    end,
                    winner,
                    tuple(sorted(contexts, key=lambda item: item.evidence_id)),
                )
            )
        else:
            covering = [
                item for item in context_items if item.start < end and item.end > start
            ]
            if covering:
                output.append(_evidence_only_slice(start, end, covering))

    return _merge_identical_touching(tuple(output))

File: packages/daytrace-core/src/fusion.ts
import type {
  ActivitySlice,
  ContextSignal,
  DiagnosticCode,
  SanitizedObservation,
} from "./models.js";
import { timestampMs } from "./time.js";

const BROWSER_APPS = ["edge", "chrome", "firefox", "brave", "opera", "safari"] as const;
const EDITOR_APPS = ["code", "codium", "pycharm", "idea", "sublime", "zed"] as const;

function appMatches(app: string | undefined, families: readonly string[]): boolean {
  const folded = (app ?? "").toLocaleLowerCase();
  return families.some((name) => folded.includes(name));
}

function context(item: SanitizedObservation): ContextSignal {
  return {
    kind: item.kind,
    evidenceId: item.evidenceId,
    ...(item.title === undefined ? {} : { title: item.title }),
    ...(item.project === undefined ? {} : { project: item.project }),
    ...(item.file === undefined ? {} : { file: item.file }),
    ...(item.urlHost === undefined ? {} : { urlHost: item.urlHost }),
    ...(item.urlPath === undefined ? {} : { urlPath: item.urlPath }),
    ...(item.language === undefined ? {} : { language: item.language }),
  };
}

function contextKey(item: ContextSignal): string {
  return JSON.stringify(item);
}

function sameContexts(left: readonly ContextSignal[], right: readonly ContextSignal[]): boolean {
  return left.length === right.length && left.every((item, index) => contextKey(item) === contextKey(right[index] as ContextSignal));
}

function mergeIdenticalTouching(slices: readonly ActivitySlice[]): readonly ActivitySlice[] {
  const merged: ActivitySlice[] = [];
  for (const item of slices) {
    const previous = merged.at(-1);
    if (
      previous !== undefined
      && previous.end === item.start
      && previous.focused === item.focused
      && previous.app === item.app
      && previous.title === item.title
      && sameContexts(previous.contexts, item.contexts)
    ) {
      merged[merged.length - 1] = {
        ...previous,
        end: item.end,
        evidenceIds: [...new Set([...previous.evidenceIds, ...item.evidenceIds])].sort(),
      };
    } else {
      merged.push(item);
    }
  }
  return merged;
}

export function fuseObservations(
  observations: readonly SanitizedObservation[],
  diagnose: (code: DiagnosticCode) => void,
): readonly ActivitySlice[] {
  const items = [...observations].sort(
    (left, right) => timestampMs(left.start) - timestampMs(right.start)
      || left.evidenceId.localeCompare(right.evidenceId),
  );
  const windows = items.filter((item) => item.kind === "current-window");
  const contextItems = items.filter((item) => item.kind !== "current-window");
  const boundaries = [...new Set(items.flatMap((item) => [timestampMs(item.start), timestampMs(item.end)]))]
    .sort((left, right) => left - right);
  const output: ActivitySlice[] = [];
  for (let index = 0; index < boundaries.length - 1; index += 1) {
    const start = boundaries[index];
    const end = boundaries[index + 1];
    if (start === undefined || end === undefined) continue;
    const coveringWindows = windows.filter(
      (item) => timestampMs(item.start) < end && timestampMs(item.end) > start,
    );
    if (coveringWindows.length > 0) {
      if (coveringWindows.length > 1) diagnose("window-conflict");
      const winner = [...coveringWindows].sort((left, right) => left.evidenceId.localeCompare(right.evidenceId))[0];
      if (winner === undefined) continue;
      const contexts = contextItems.filter((item) => {
        if (timestampMs(item.start) >= end || timestampMs(item.end) <= start) return false;
        return (item.kind === "browser" && appMatches(winner.app, BROWSER_APPS))
          || (item.kind === "editor" && appMatches(winner.app, EDITOR_APPS));
      }).map(context).sort((left, right) => left.evidenceId.localeCompare(right.evidenceId));
      output.push({
        start: new Date(start).toISOString(),
        end: new Date(end).toISOString(),
        focused: true,
        ...(winner.app === undefined ? {} : { app: winner.app }),
        ...(winner.title === undefined ? {} : { title: winner.title }),
        contexts,
        evidenceIds: [...new Set([winner.evidenceId, ...contexts.map((item) => item.evidenceId)])].sort(),
      });
      continue;
    }
    const covering = contextItems.filter(
      (item) => timestampMs(item.start) < end && timestampMs(item.end) > start,
    );
    if (covering.length > 0) {
      const contexts = covering.map(context).sort((left, right) => left.evidenceId.localeCompare(right.evidenceId));
      output.push({
        start: new Date(start).toISOString(),
        end: new Date(end).toISOString(),
        focused: false,
        contexts,
        evidenceIds: contexts.map((item) => item.evidenceId),
      });
    }
  }
  return mergeIdenticalTouching(output);
}
